src/undocommand.js

```diff
@@ -0,0 +1,200 @@
+			// After the range is broken into two ranges, we have to transform both of those ranges separately.
+			// For that reason, we keep all transformed ranges in one array and operate on it.
+			let transformed = [ originalRange ];
+
+			for ( let delta of deltas ) {
+				for ( let operation of delta.operations ) {
+					// We look through all operations from all deltas.
+
+					for ( let t = 0; t < transformed.length; t++ ) {
+						// We transform every range by every operation.
+						// We keep current state of transformation in `transformed` array and update it.
+						let result;
+
+						switch ( operation.type ) {
+							case 'insert':
+								result = transformed[ t ].getTransformedByInsertion( operation.position, operation.nodeList.length, true );
+								break;
+
+							case 'move':
+							case 'remove':
+							case 'reinsert':
+								result = transformed[ t ].getTransformedByMove( operation.sourcePosition, operation.targetPosition, operation.howMany, true );
+								break;
+						}
+
+						// If we have a transformation result, we substitute it in `transformed` array with
+						// the range that got transformed. Keep in mind that the result is an array
+						// and may contain multiple ranges.
+						if ( result ) {
+							transformed.splice( t, 1, ...result );
+
+							// Fix iterator.
+							t = t + result.length - 1;
+						}
+					}
+				}
+			}
+
+			// After `originalRange` got transformed, we have an array of ranges. Some of those
+			// ranges may be "touching" -- they can be next to each other and could be merged.
+			// Let's do this. First, we have to sort those ranges because they don't have to be
+			// in an order.
+			transformed.sort( ( a, b ) => a.start.isBefore( b.start ) ? -1 : 1 );
+
+			// Then we check if two consecutive ranges are touching. We can do it pair by pair
+			// in one dimensional loop because ranges are sorted.
+			for ( let i = 1 ; i < transformed.length; i++ ) {
+				let a = transformed[ i - 1 ];
+				let b = transformed[ i ];
+
+				if ( a.end.isTouching( b.start ) ) {
+					a.end = b.end;
+					transformed.splice( i, 1 );
+					i--;
+				}
+			}
+
+			// For each `originalRange` from `ranges`, we take only one transformed range.
+			// This is because we want to prevent situation where single-range selection
+			// got transformed to mulit-range selection. We will take the first range that
+			// is not in the graveyard.
+			const transformedRange = transformed.find(
+				( range ) => range.start.root != this.editor.document.graveyard
+			);
+
+			if ( transformedRange ) {
+				transformedRanges.push( transformedRange );
+			}
+		}
+
+		// `transformedRanges` may be empty if all ranges ended up in graveyard.
+		// If that is the case, do not restore selection.
+		if ( transformedRanges.length ) {
+			this.editor.document.selection.setRanges( transformedRanges, selectionState.isBackward );
+		}
+
+		this.fire( 'undo', undoBatch );
```

Comment:
It's a bit strange that the undo command fires the `undo` event. Especially when it's `redoCommand.on( 'undo', ... )`. So maybe we could find some better name. Like `revertedBatch`?